### ISSUE TITLE
Win2d graphics erase after GPU device re-enabled

### ISSUE THREAD

**Issue by johniskew** [microsoft/Win2D#750]
I am utilizing Win2d in a UWP app on Windows 10, and am manually handling lost devices. I am subscribing to the CanvasDevice.DeviceLost event, like so:

In my constructor:

CanvasDevice = CanvasDevice.GetSharedDevice();
CanvasDevice.DeviceLost += CanvasDevice_DeviceLost;
DeviceLost handler:

public async void CanvasDevice_DeviceLost(CanvasDevice sender, object args)
{
    CanvasDevice = CanvasDevice.GetSharedDevice();
    await LoadOriginalBitmap();
}

To test whether DeviceLost is firing correctly, I am using the advice here: https://microsoft.github.io/Win2D/html/HandlingDeviceLost.htm

This works relatively well when I disable the display adapter - DeviceLost fires and I re-create the device and redraw my graphics. I would assume it is using software rendering at this point?

The problem is, if I RE_ENABLE the display adapter, my Win2d graphics get erased as I assume the CanvasDevice switches back to GPU rendering. BUT, there is no DeviceFound or similar equivalent to handle this (DeviceLost doesn't fire in this case). What can I hook into to redraw things?

I wanted to note, that if I do everything within a CanvasControl Draw event, things seem to fire off properly. But forcing the design of the code I'm working with into that convention is very convoluted.  I need to handle the DeviceLost exceptions manually.

**Comment by johniskew** [comment]
Here is a way to reproduce this problem:

1) I disable both GPUs: (a) Intel HD Graphics 5500, (2) NVIDIA Geforce 840M. (I did this to test how robust the error handling will be)

2) Run the app, Click button.  Image draws fine.

3) Enable Intel GPU ONLY, graphic disappears, click button, throws exception as it should:
Exception thrown: 'System.Exception' in App1.exe
WinRT information: The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action.

CAUGHT EXCEPTION
IN DEVICELOST
IN DEVICELOST <-- Why is this firing twice????

4) Click button again, image draws, all good.

5) Disable Intel, graphic disappears, click button, throws exception:
CAUGHT EXCEPTION
Exception thrown: 'System.Exception' in App1.exe
WinRT information: The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action.

However, RaiseDeviceLost is NOT firing, DeviceLost does not get called, and I can't reset the CanvasDevice!

**Comment by shawnhar** [comment]
I see a couple of things wrong with this code:

1) You are mapping _all_ exceptions to the RaiseDeviceLost handler, rather than only device lost exceptions.  You need to filter the exception handler to only cases where canvasDevice.IsDeviceLost(e.ErrorCode) returns true, as shown at https://github.com/microsoft/Win2D/issues/750

2) You subscribe to the DeviceLost event in your constructor, but then in the DeviceLost handler you change MyDevice to a different instance.  At this point you are only subscribed to the DeviceLost event of the original device instance, but not the new one.

**Comment by johniskew** [comment]
Thanks so much for answering so quickly! For (1), absolutely, I just wanted to see what else might be thrown there.

I've updated code for (2) to be:

        private void CanvasDevice_DeviceLost(CanvasDevice sender, object args)
        {
            Debug.WriteLine("IN DEVICELOST");
            MyDevice.DeviceLost -= CanvasDevice_DeviceLost;
            MyDevice = CanvasDevice.GetSharedDevice();
            MyDevice.DeviceLost += CanvasDevice_DeviceLost;
        }

Works well. I noticed if I did not remove the previous handler, then the DeviceLost will get called in growing multiples. Good lesson for me there. Maybe consider adding a fuller example like this on the "Manual LostDevice" article.

Two quick questions:
(a) If I force a software render device like: CanvasDevice.GetSharedDevice(true). Is there any possibility of getting a LostDevice? Or would LostDevice just be for a lost GPU? My app really isn't graphic intensive, so I was hoping to get rid of the need to possibly renew the device.
(b) It is mentioned at this example [https://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_UI_Xaml_CanvasImageSource_Recreate.htm] to handle XAML's CompositionTarget.SurfaceContentsLost. Would it be necessary to track this as well as CanvasDevice.DeviceLost? Or will DeviceLost fire automatically if SurfaceContentsLost is firing? Or vice-versa.

Thanks again for your help and speed - a real credit to you guys!!
